File: backend/workflow/vm_loop.go
package workflow

import (
	"context"
	"errors"
	"sync"

	"github.com/dop251/goja"
)

type vmLoop struct {
	vm       *goja.Runtime
	dispatch chan func()
	wg       sync.WaitGroup
}

func newVMLoop(vm *goja.Runtime) *vmLoop {
	return &vmLoop{vm: vm, dispatch: make(chan func(), 1024)}
}

func (l *vmLoop) async(ctx context.Context, work func() (any, error)) goja.Value {
	promise, resolve, reject := l.vm.NewPromise()
	l.wg.Add(1)
	go func() {
		defer l.wg.Done()
		value, err := work()
		select {
		case <-ctx.Done():
			err = ctx.Err()
		default:
		}
		select {
		case l.dispatch <- func() {
			if err != nil {
				_ = reject(err.Error())
			} else {
				_ = resolve(jsonShape(value))
			}
		}:
		case <-ctx.Done():
		}
	}()
	return l.vm.ToValue(promise)
}

func (l *vmLoop) wait() {
	l.wg.Wait()
}

func (l *vmLoop) await(ctx context.Context, value goja.Value) (goja.Value, error) {
	if value == nil {
		return goja.Undefined(), nil
	}
	promise, ok := value.Export().(*goja.Promise)
	if !ok {
		return value, nil
	}
	for promise.State() == goja.PromiseStatePending {
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case fn := <-l.dispatch:
			fn()
		}
	}
	switch promise.State() {
	case goja.PromiseStateFulfilled:
		return promise.Result(), nil
	case goja.PromiseStateRejected:
		return nil, errors.New(promise.Result().String())
	default:
		return nil, errors.New("promise remained pending")
	}
}

File: backend/workflow/vm_loop.ts
// PORT: backend/workflow/vm_loop.go

export class vmLoop {
  private activePromisesCount = 0;

  async(ctx: AbortSignal, work: () => Promise<any>): Promise<any> {
    this.activePromisesCount++;
    return new Promise((resolve, reject) => {
      let settled = false;
      const onAbort = () => {
        if (settled) return;
        settled = true;
        this.activePromisesCount--;
        reject(new Error("interrupted"));
      };
      if (ctx.aborted) {
        onAbort();
        return;
      }
      ctx.addEventListener("abort", onAbort);

      work()
        .then((val) => {
          if (settled) return;
          settled = true;
          ctx.removeEventListener("abort", onAbort);
          this.activePromisesCount--;
          resolve(val);
        })
        .catch((err) => {
          if (settled) return;
          settled = true;
          ctx.removeEventListener("abort", onAbort);
          this.activePromisesCount--;
          reject(err);
        });
    });
  }

  wait(): void {
    // wait handles standard Node/Bun thread settle which is native.
  }

  async await(ctx: AbortSignal, value: any): Promise<any> {
    if (value === undefined || value === null) {
      return undefined;
    }
    if (typeof value.then === "function") {
      if (ctx.aborted) {
        throw new Error("interrupted");
      }
      return new Promise((resolve, reject) => {
        let settled = false;
        const onAbort = () => {
          if (settled) return;
          settled = true;
          reject(new Error("interrupted"));
        };
        ctx.addEventListener("abort", onAbort);
        value
          .then((res: any) => {
            if (settled) return;
            settled = true;
            ctx.removeEventListener("abort", onAbort);
            resolve(res);
          })
          .catch((err: any) => {
            if (settled) return;
            settled = true;
            ctx.removeEventListener("abort", onAbort);
            reject(err);
          });
      });
    }
    return value;
  }
}

export function newVMLoop(): vmLoop {
  return new vmLoop();
}

/*
PORT STATUS
source path: backend/workflow/vm_loop.go
source lines: 75
draft lines: 55
confidence: high
status: phase_b_compile
*/
